src/idpi/grib_decoder.py

```diff
@@ -82,6 +52,171 @@ def _extract_pv(pv):
     }
 
 
+@dc.dataclass
+class _FieldBuffer:
+    dims: tuple[str, ...] | None = None
+    hcoords: dict[str, xr.DataArray] = dc.field(default_factory=dict)
+    metadata: dict[str, typing.Any] = dc.field(default_factory=dict)
+    time_meta: dict[int, dict] = dc.field(default_factory=dict)
+    values: dict[tuple[int, ...], np.ndarray] = dc.field(default_factory=dict)
+
+    def load(self, field):
+        dim_keys = (
+            ("perturbationNumber", "step", "level")
+            if _is_ensemble(field)
+            else ("step", "level")
+        )
+        key = field.metadata(*dim_keys)
+        logger.debug("Received field for key: %s", key)
+        self.values[key] = field.to_numpy(dtype=np.float32)
+
+        step = key[-2]  # assume all members share the same time steps
+        if step not in self.time_meta:
+            self.time_meta[step] = field.metadata(namespace="time")
+
+        if not self.dims:
+            self.dims = tuple(DIM_MAP[d] for d in dim_keys) + ("y", "x")
+
+        if not self.metadata:
+            self.metadata = {
+                "message": field.message(),
+                **metadata.extract(field.metadata()),
+            }
+
+        if not self.hcoords:
+            self.hcoords = {
+                dim: xr.DataArray(dims=("y", "x"), data=values)
+                for dim, values in field.to_latlon().items()
+            }
+
+    def _gather_coords(self):
+        if self.dims is None:
+            raise RuntimeError("No dims.")
+
+        coord_values = zip(*self.values)
+        unique = (sorted(set(values)) for values in coord_values)
+        coords = {dim: c for dim, c in zip(self.dims[:-2], unique)}
+
+        if missing := [
+            combination
+            for combination in product(*coords.values())
+            if combination not in self.values
+        ]:
+            msg = f"Missing combinations: {missing}"
```

Comment:
I believe this can only happen if FDB skips a grib message or the underlying archive is incomplete. This exception is raised when there are missing combinations of coordinate values based on the other messages that have been received. There's no inspection of the user's request in this instance. I think the message does not necessarily need to be user friendly accounting for the fact that it does not result of user error.

Reply:
The underlying FDB could be incomplete. So all GRIB message recieved should have the same coordinates?